crates/sui-adapter/src/adapter.rs

```diff
@@ -1071,19 +1074,54 @@ fn serialize_object<'a>(
                 error,
             ));
         }
-        InputObjectKind::SharedMoveObject { .. } if !object.is_shared() => {
-            let error = format!(
-                "Argument at index {} populated with an immutable or owned object id {} \
-                        but an shared object was expected",
-                idx, object_id
-            );
-            return Err(ExecutionError::new_with_source(
-                ExecutionErrorKind::entry_argument_error(
-                    idx,
-                    EntryArgumentErrorKind::ObjectKindMismatch,
-                ),
-                error,
-            ));
+        InputObjectKind::SharedMoveObject { mutable, .. } => {
```

Comment:
@amnn - wanted to confirm if this is the right place and way to enforce that mutability of InputObjectKind matches mutability on the move call

Reply:
How bad is it if we don't specicy it here, but when we schedule it, we load the function interface from the package and see if the argument is mutable or not?

Reply:
My only concern is that it might get complicated having lots of package loading all over the place. Though, do we already load the package when scheduling?

Reply:
But I'm not sure how the scheduler would insert this information?

Reply:
I thought little bit about it - I don't think this will be possible without making things very complex.

There is no guarantee that by the time consensus handler receives transaction the move module will be available - it is a job of transaction manager to wait for dependencies. Fact that we have a certificate in consensus means that majority of nodes have a move module and it will eventually be available to our node, but nothing guarantees that at the very moment we receive it from consensus we have a move module for transaction

Reply:
Possibly dumb question -- can we shift this even earlier than scheduling, i.e. to certificate signing?  If not, this seems like the right place for it @andll.

Reply:
This should be doable but it will have performance and scalability implications. Given that low latency for owned object transactions is one of the key selling point of sui, I am a bit hesitant to add work there, if we have alternatives like generating value of mutable field in SDK like you mentioned below
But I am also not sure how hard will it be to generate this in SDK

Reply:
That's a reasonable take, especially in light of the ability for SDKs to eventually infer this information.